src/mlpack/core/cv/k_fold_cv.hpp

```diff
@@ -0,0 +1,219 @@
+ * supposed to be passed after the k parameter in the KFoldCV constructor.
+ *
+ * For example, you can run 10-fold cross-validation for SoftmaxRegression in
+ * the following way.
+ *
+ * @code
+ * // 100-point 5-dimensional random dataset.
+ * arma::mat data = arma::randu<arma::mat>(5, 100);
+ * // Random labels in the [0, 4] interval.
+ * arma::Row<size_t> labels =
+ *     arma::randi<arma::Row<size_t>>(100, arma::distr_param(0, 4));
+ * size_t numClasses = 5;
+ *
+ * KFoldCV<SoftmaxRegression<>, Accuracy> cv(10, data, labels, numClasses);
+ *
+ * double lambda = 0.1;
+ * double softmaxAccuracy = cv.Evaluate(lambda);
+ * @endcode
+ *
+ * @tparam MLAlgorithm A machine learning algorithm.
+ * @tparam Metric A metric to assess the quality of a trained model.
+ * @tparam MatType The type of data.
+ * @tparam PredictionsType The type of predictions (should be passed when the
+ *     predictions type is a template parameter in Train methods of
+ *     MLAlgorithm).
+ * @tparam WeightsType The type of weights (should be passed when weighted
+ *     learning is supported, and the weights type is a template parameter in
+ *     Train methods of MLAlgorithm).
+ */
+template<typename MLAlgorithm,
+         typename Metric,
+         typename MatType = arma::mat,
+         typename PredictionsType =
+             typename MetaInfoExtractor<MLAlgorithm, MatType>::PredictionsType,
+         typename WeightsType =
+             typename MetaInfoExtractor<MLAlgorithm, MatType,
+                 PredictionsType>::WeightsType>
+class KFoldCV :
+    private CVBase<MLAlgorithm, MatType, PredictionsType, WeightsType>
+{
+ public:
+  /**
+   * Construct an object for running k-fold cross-validation.
+   *
+   * @param k Number of folds (should be at least 2).
+   * @param args Basic constructor arguments for MLAlgortithm (see the CVBase
+   *     constructors for reference).
+   */
+  template<typename... CVBaseArgs>
+  KFoldCV(const size_t k, const CVBaseArgs&... args);
+
+  /**
+   * Run k-fold cross-validation.
+   *
+   * @param args Arguments for MLAlgorithm (in addition to the passed
+   *     ones in the constructor).
+   */
+  template<typename... MLAlgorithmArgs>
+  double Evaluate(const MLAlgorithmArgs& ...args);
+
+  //! Access and modify a model from the last run of k-fold cross-validation.
+  MLAlgorithm& Model();
+
+ private:
+  //! A short alias for CVBase.
+  using Base = CVBase<MLAlgorithm, MatType, PredictionsType, WeightsType>;
+
+  //! The number of bins in the dataset.
+  const size_t k;
+
+  /**
+   * Variables for storing the extended (by repeating the first k - 2 bins)
+   * dataset.
+   */
+  MatType xs;
+  PredictionsType ys;
+  WeightsType weights;
+
+  //! The size of each bin in terms of data points.
+  size_t binSize;
+
+  //! The size of each training subset in terms of data points.
+  size_t trainingSubsetSize;
+
+  //! A pointer to a model from the last run of k-fold cross-validation.
+  std::unique_ptr<MLAlgorithm> modelPtr;
```

Comment:
I forgot to follow up with this on the last simple_cv PR, and we're about to
merge that so I don't want to bring this up again there.  Nowhere else in the
mlpack codebase is std::unique_ptr<> used, so for consistency I'd prefer to
avoid the use of the class.

I see a couple of possibilities---you can use a bare pointer or actually hold an
MLAlgorithm internally and I think this will not affect the external API you
have already provided.  At a quick glance, I think that both of those should be
possible; but perhaps there is a catch I have overlooked.

This is a minor comment that doesn't really affect the actual functionality of
the CV code at all (I think), so we can consider it low-priority.  If you
prefer, I can submit the PR to make these changes, but it will be a few weeks
until that happens.  (Maybe I will prepare it on the flight home, but that will
be August 12...)